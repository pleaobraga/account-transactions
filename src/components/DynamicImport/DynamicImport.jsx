import React, { Component } from 'react'
import PropTypes from 'prop-types'

class DynamicImport extends Component {
  state = {
    component: null,
  }

  componentDidMount() {
    const { loadComponent, ErrorComponent } = this.props

    loadComponent()
      .then((comp) => {
        comp.default
          ? this.setState({ Component: comp.default })
          : this.setState({ Component: ErrorComponent })
      })
      .catch(() => this.setState({ Component: ErrorComponent }))
  }

  render() {
    const { LoadingComponent } = this.props
    const { Component } = this.state

    return Component ? <Component /> : <LoadingComponent />
  }
}

DynamicImport.defaultProps = {
  ErrorComponent: () => <div id="error">Error</div>,
  LoadingComponent: () => <div id="loading">Loading...</div>,
}

DynamicImport.propTypes = {
  loadComponent: PropTypes.func.isRequired,
  LoadingComponent: PropTypes.func,
  ErrorComponent: PropTypes.func,
}

export default DynamicImport
