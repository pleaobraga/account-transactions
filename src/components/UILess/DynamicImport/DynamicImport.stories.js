import React from 'react'
import { storiesOf } from '@storybook/react'
import DynamicImport from './DynamicImport'

export default storiesOf('Components | DynamicImport', module).add(
  'default',
  () => <DynamicImport loadComponent={() => Promise.resolve('data')} />,
  { info: { inline: true, header: false } }
)
