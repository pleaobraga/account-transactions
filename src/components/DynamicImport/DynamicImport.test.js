import 'babel-polyfill'
import React from 'react'
import { mount } from 'enzyme'
import DynamicImport from './DynamicImport'

const a = () => <div id="content">test</div>

const request = data => {
  return new Promise(resolve => {
    data ? resolve({ default: a }) : resolve({ default: null })
  })
}

describe.only('DynamicImport', () => {
  it('render proprely', () => {
    const dynamicImport = mount(
      <DynamicImport loadComponent={() => request('test')} />
    )

    expect(dynamicImport).toMatchSnapshot()
    expect(dynamicImport.find('#loading').length).toBe(1)
  })

  it('Should render component', async () => {
    const dynamicImport = await mount(
      <DynamicImport loadComponent={() => request('test')} />
    )

    dynamicImport.update()

    expect(dynamicImport).toMatchSnapshot()
    expect(dynamicImport.find('#content').length).toBe(1)
  })

  it('Should render Error component', async () => {
    const dynamicImport = await mount(
      <DynamicImport loadComponent={() => request()} />
    )

    dynamicImport.update()

    expect(dynamicImport).toMatchSnapshot()
    expect(dynamicImport.find('#error').length).toBe(1)
  })
})
