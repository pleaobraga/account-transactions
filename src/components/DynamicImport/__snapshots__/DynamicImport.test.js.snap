// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicImport Should render Error component 1`] = `
<DynamicImport
  ErrorComponent={[Function]}
  LoadingComponent={[Function]}
  loadComponent={[Function]}
>
  <ErrorComponent>
    <div
      id="error"
    >
      Error
    </div>
  </ErrorComponent>
</DynamicImport>
`;

exports[`DynamicImport Should render component 1`] = `
<DynamicImport
  ErrorComponent={[Function]}
  LoadingComponent={[Function]}
  loadComponent={[Function]}
>
  <a>
    <div
      id="content"
    >
      test
    </div>
  </a>
</DynamicImport>
`;

exports[`DynamicImport render proprely 1`] = `
<DynamicImport
  ErrorComponent={[Function]}
  LoadingComponent={[Function]}
  loadComponent={[Function]}
>
  <LoadingComponent>
    <div
      id="loading"
    >
      Loading...
    </div>
  </LoadingComponent>
</DynamicImport>
`;
